### DISCUSSION TITLE
Generate for fastcdr

### DISCUSSION THREAD

**Discussion by MichaelFuchs123** [eProsima/Fast-DDS-Gen#304] [Q&A]
I am only able to generate IDL definition/implementation files with fastddsgen that use the `eprosima::fastcdr:Cdr` class.  However, I would also like to be able to generate code for those same IDL files but that use `eprosima::fastcdr:FastCdr` class instead.  

- This link seems to imply that it is possible: https://www.eprosima.com/docs/fast-buffers/0.3.0/html/group___f_o_o_f_a_s_t_c_d_r.html
- I can also see in the actual source code for fastddsgen, that it also seems possible to generate this based on, for example, the following snippet. 
  * File link: https://github.com/eProsima/Fast-DDS-Gen/blob/master/src/main/java/com/eprosima/fastcdr/idl/templates/TypesSource.stg
  * Lines 239-255:

```
$if(ctx.fastcdr)$
void $struct.name$::serialize(
        eprosima::fastcdr::FastCdr &scdr) const
{
    $if(struct.inheritance)$$struct.inheritance.scopedname$::serialize(scdr);$endif$

    $struct.members : { member |$if(!member.annotationNonSerialized)$$object_serialization(ctx=ctx, object=member)$$endif$}; separator=""$
}

void $struct.name$::deserialize(
        eprosima::fastcdr::FastCdr& dcdr)
{
    $if(struct.inheritance)$$struct.inheritance.scopedname$::deserialize(dcdr);$endif$

    $struct.members : { member |$if(!member.annotationNonSerialized)$$object_deserialization(ctx=ctx, object=member)$$endif$}; separator="\n"$
}
$endif$
```

**Comment by sbrett9** [comment]
I also need the same thing.
There is an option in fastddsgen. -fusion.
If that is enabled, then the code seems to generate with FastCdr support.

However there seems to be a few issues.
First, when using fastddsgen option -cdr v2 the code that is generated seems to be not quite updated perhaps for some changes to FastCdr 2.2.1 (which is what I was using), and you get errors attributed to the MemberId class apparently not having a member named "serialize", nor any overload of stream << & >> operators.
These do exist for MemberId, but they're part of the XCdr extensions  defined in Cdr.h, not FastCdr.h

Second, when -fusion is enabled, sometimes the code generated has unicode characters output in it.
Clearly a bug in FastDDSGen perhaps attributed to the -fusion option not even being published. Perhaps something hasn't been maintained over time.
I am not sure if this is an issue with the .stg files or something in the java that uses the -fusion flag as an input.
I think it has something to do with the random name being generated for variables in for loops or other places and the counter not quite checking if they've reached the end of the ascii character set and should start adding more characters.

So, if you use v1 and then are able to fix the variable name fabrication ascii overrun issue, which I think is Context.java, then it might just work!
